stan/math/prim/mat/fun/gp_matern52_cov.hpp

```diff
@@ -0,0 +1,323 @@
+#ifndef STAN_MATH_PRIM_MAT_FUN_GP_MATERN52_COV_HPP
+#define STAN_MATH_PRIM_MAT_FUN_GP_MATERN52_COV_HPP
+
+#include <stan/math/prim/mat/fun/Eigen.hpp>
+#include <stan/math/prim/mat/fun/sqrt.hpp>
+#include <stan/math/prim/scal/err/check_finite.hpp>
+#include <stan/math/prim/scal/err/check_nonnegative.hpp>
+#include <stan/math/prim/scal/err/check_not_nan.hpp>
+#include <stan/math/prim/scal/err/check_positive_finite.hpp>
+#include <stan/math/prim/scal/fun/square.hpp>
+#include <stan/math/prim/scal/fun/squared_distance.hpp>
+#include <stan/math/prim/scal/meta/return_type.hpp>
+#include <stan/math/prim/scal/fun/divide.hpp>
+#include <cmath>
+#include <vector>
+
+namespace stan {
+namespace math {
+
+/**
+ * Returns a Matern 5/2 covariance matrix with one input vector
+ *
+ * \f[ k(x, x') = \sigma^2\bigg(1 +
+ *  \frac{\sqrt{5}d(x, x')}{l} + \frac{5d(x, x')^2}{3l^2}\bigg)
+ *   exp\bigg(-\frac{5 d(x, x')}{l}\bigg)
+ * \f]
+ *
+ * where \f$ d(x, x') \f$ is the Euclidean distance.
+ *
+ * @tparam T_x type of elements contained in vector x
+ * @tparam type of element of sigma, marginal standard deviation
+ * @tparam T_l type of elements of length scale
+ *
+ * @param x std::vector of elements that can be used in stan::math::distance
+ * @param length_scale length scale
+ * @param sigma standard deviation that can be used in stan::math::square
+ * @throw std::domain error if sigma <= 0, l <= 0, or x is nan or inf
+ *
+ */
+template <typename T_x, typename T_s, typename T_l>
+inline typename Eigen::Matrix<typename return_type<T_x, T_s, T_l>::type,
+                              Eigen::Dynamic, Eigen::Dynamic>
+gp_matern52_cov(const std::vector<T_x> &x, const T_s &sigma,
+                const T_l &length_scale) {
+  using std::exp;
+  using std::pow;
+
+  size_t x_size = x.size();
+  for (size_t n = 0; n < x_size; ++n)
+    check_not_nan("gp_matern52_cov", "x", x[n]);
+
+  check_positive_finite("gp_matern52_cov", "marginal variance", sigma);
+  check_positive_finite("gp_matern52_cov", "length scale", length_scale);
+
+  Eigen::Matrix<typename return_type<T_x, T_s, T_l>::type, Eigen::Dynamic,
+                Eigen::Dynamic>
+      cov(x_size, x_size);
+
+  if (x_size == 0)
+    return cov;
+
+  T_s sigma_sq = square(sigma);
+  T_l root_5_inv_l = pow(5.0, 0.5) / length_scale;
+  T_l inv_l_sq_5_3 = 5.0 / (3.0 * square(length_scale));
+  T_l neg_root_5_inv_l = -1.0 * pow(5.0, 0.5) / length_scale;
+  typename return_type<T_x>::type sq_distance;
```

Comment:
I think there's a more appropriate template meta program to use here. I think `scalar_type<T_x>::type` is the right one.